onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu

```diff
@@ -139,129 +142,475 @@
+    ConcatKVInPlaceLarge<float2><<<grid, block, 0, stream>>>(past_sequence_length,
+                                                             present_sequence_length,
+                                                             H,
+                                                             reinterpret_cast<float2*>(data.present_key),
+                                                             reinterpret_cast<const float2*>(data.key),
+                                                             past_kv_format == AttentionQkvFormat::Q_K_V_BSNH);
+    ConcatKVInPlaceLarge<float2><<<grid, block, 0, stream>>>(past_sequence_length,
+                                                             present_sequence_length,
+                                                             H,
+                                                             reinterpret_cast<float2*>(data.present_value),
+                                                             reinterpret_cast<const float2*>(data.value),
+                                                             past_kv_format == AttentionQkvFormat::Q_K_V_BSNH);
+  }
+  return CUDA_CALL(cudaGetLastError());
+}
+
+// Kernel for use with memory efficient kernel... kv_in is grouped and of bnsh or bsnh... kv_out is ungrouped and bsnh
+template <typename T>
+__global__ void UngroupAndTranspose(const T* kv_in,
+                                    T* kv_out,
+                                    const int in_seqlen,
+                                    const int kv_num_heads,
+                                    const bool is_bsnh) {
+  const int h = threadIdx.x;
+  const int out_n = threadIdx.y;
+  const int s = blockIdx.x;
+  const int b = blockIdx.y;
+
+  const int out_seqlen = gridDim.x;
+  const int q_num_heads = blockDim.y;
+  const int H = blockDim.x;
+
+  const int q_kv_head_ratio = q_num_heads / kv_num_heads;
+  const int out_batch_stride = out_seqlen * q_num_heads * H;
+  const int out_row_stride = q_num_heads * H;
+  const int out_head_stride = H;
+
+  const int in_batch_stride = in_seqlen * kv_num_heads * H;
+  const int in_row_stride = is_bsnh ? kv_num_heads * H : H;
+  const int in_head_stride = is_bsnh ? H : in_seqlen * H;
+  const int in_n = out_n / q_kv_head_ratio;
+
+  const int out_offset = out_batch_stride * b + out_row_stride * s + out_head_stride * out_n + h;
+  const int in_offset = in_batch_stride * b + in_row_stride * s + in_head_stride * in_n + h;
+  kv_out[out_offset] = kv_in[in_offset];
+}
+
+template <typename T>
+__global__ void UngroupAndTransposeLarge(const T* kv_in,
+                                         T* kv_out,
+                                         const int H,
+                                         const int in_seqlen,
+                                         const int kv_num_heads,
+                                         const bool is_bsnh) {
+  int h = threadIdx.x;
+  const int out_n = threadIdx.y;
```

Comment:
The memory access is not continuous with this 2D implementation, even inside a warp. Take Llama2 13B model for example, its hidden size is 5120 with num of head 40 and head size 128.  H will be 32 (128/4) and H * number_of_head = 1280 > max_thread_per_block(1024), so the thread block will be (25, 40). Here we use 25 threads to handle 32 values and it is inefficient. It is better to just implement it with 1D thread block, for example with 256 threads. And we use 256 threads to handle 1280 items. writing will be continuous for whole thread block and read will be continuous for a warp. It can be extended to other cases.